### PR TITLE
consolidate koop, lib, app

### PR DESCRIPTION
**Pull request by ngoldman**
This was a lot easier than I thought it would be.

I feel like we need to test this more thoroughly since a lot of providers use the insides of koop in weird ways, but all tests are passing, so...

@dmfenton thoughts?

### CODE REVIEW

File: lib/Cache.js [hunk 1]
```diff
@@ -1,4 +1,9 @@
-function Cache (koop) {
+/**
+ * Cache constructor.
+ * Exposes methods for managing the underlying database.
+ * @class
+ */
+function Cache () {
```
Comment (ngoldman) [inline lib/Cache.js]: Turns out koop is never used internally in this file.

File: lib/BaseModel.js [hunk 2]
```diff
@@ -4,8 +4,9 @@ var spatialReference = require('./SpatialReference')
 /**
  * Uses the module pattern for clean inheritance in models
  * exposes shared functionality across providers, typically things that require central code in koop
+ * TODO: refactor so that koop app isn't passed in at all
  *
- * @param {Object} koop
+ * @param {Object} koop - instance of koop middleware app
  */
 function Model (koop) {
```
Comment (ngoldman) [inline lib/BaseModel.js]: Made a note above -- in a perfect future we're not giving the model constructor a copy of the entire koop app.

File: index.js [hunk 3]
```diff
@@ -3,32 +3,34 @@ var bodyParser = require('body-parser')
...
-   */
+  // inherit everything from lib ¯\_(ツ)_/¯
+  for (var method in lib) {
+    if (lib.hasOwnProperty(method)) koop[method] = lib[method]
+  }
 
-  koop.config = config = config || {}
-  koop.log = new koop.Logger(config)
+  koop.config = config || {}
+  koop.log = new koop.Logger(koop.config)
```
Comment (dmfenton) [inline index.js]: this just seems really weird.
we set a property called log by constructing an object called logger.
is there a better way?

File: index.js [hunk 4]
```diff
@@ -3,32 +3,34 @@ var bodyParser = require('body-parser')
...
 
   // init koop centralized file access
   // this allows us to turn the FS access off globally
   koop.files = new koop.Files({
-    config: config,
+    config: koop.config,
     log: koop.log
   })
```
Comment (dmfenton) [inline index.js]: same as above

File: index.js [hunk 5]
```diff
@@ -175,17 +177,17 @@ module.exports = function (config) {
...
-  app.get('/providers', function (req, res, next) {
-    res.json(app.services)
+  koop.get('/providers', function (req, res, next) {
+    res.json(koop.services)
   })
 
   // serve up a provider
-  app.get('/providers/:provider', function (req, res, next) {
-    res.json(app.services[req.params.provider])
+  koop.get('/providers/:provider', function (req, res, next) {
```
Comment (dmfenton) [inline index.js]: may not part of this PR, but can we move some of these routes to a separate module or controller? There's really just for Koop-UI as far as I can tell

File: index.js [hunk 6]
```diff
@@ -203,6 +205,7 @@ module.exports = function (config) {
 
   if (koop.config.export_workers) {
     var kue = require('kue')
+
     koop.Exporter.export_q = kue.createQueue({
```
Comment (dmfenton) [inline index.js]: also probably not as part of this PR, but it seems weird to handle this export_q stuff here

File: lib/BaseModel.js (hunk 2 above)
Comment (dmfenton) [inline lib/BaseModel.js]: definitely agree

File: lib/Cache.js (hunk 1 above)
Comment (dmfenton) [inline lib/Cache.js]: I'm not so sure about this. I think this is using `koop.db` as `this.db`

**Comment by dmfenton** [comment]
Have you tried running all this locally with a registered cache?

File: lib/Cache.js (hunk 1 above)
Comment (ngoldman) [inline lib/Cache.js]: `db` is assigned here: https://github.com/koopjs/koop/pull/238/files#diff-168726dbe96b3ce427e7fedce31bb0bcL29

`koop` is never used as a param. `this` in this case is the instance of `Cache` attached to koop.

**Comment by ngoldman** [comment]
@dmfenton probably not a bad idea to do that for the sake of covering our bases, but the API for `Cache` hasn't actually changed since the `koop` param is *never* used in there.

File: index.js (hunk 6 above)
Comment (ngoldman) [inline index.js]: agree, Exporter is used in BaseModel too. Should be pulled out into a module (koop-exporter) and used only where necessary. Should also be lowercased as it's not a class/constructor.

File: index.js (hunk 5 above)
Comment (ngoldman) [inline index.js]: I would say we should keep these here. It's pretty hard to see what a koop instance is running from the outside without these. If anything I'd like to combine this and `status` which isn't really used into something informative for any koop instance. If koop-ui consumes it that's a bonus but I think it's more useful as an API endpoint to check health, status, versions, etc.

File: index.js (hunk 3 above)
Comment (ngoldman) [inline index.js]: Would be better if Logger wasn't hanging off of koop. I was going to propose removing constructors from being exposed on the koop object in a future major version bump.

File: index.js (hunk 4 above)
Comment (ngoldman) [inline index.js]: same response as above :)

**Comment by dmfenton** [comment]
Great

File: index.js [hunk 7]
```diff
@@ -3,32 +3,34 @@ var bodyParser = require('body-parser')
...
     log: koop.log
   })
 
   // the default cache is the local in-mem cache
   // to persist data you must call registerCache with db adapter
   // use the default local cache until a DB adapter mod is registered
-  koop.Cache = new koop.DataCache(koop)
+  koop.Cache = new koop.DataCache()
   koop.Cache.db = koop.LocalDB
+  koop.Exporter = koop.Exporter
```
Comment (ngoldman) [inline index.js]: just noticed this :(

Comment (dmfenton) [inline index.js]: A good ole fashion whoopsie. I say just pull it out in a commit to master

Comment (ngoldman) [inline index.js]: removed in 3072f9939a21ff914d2664fe17f5feb4b6eb62e9

**Pull request merged**